Show warning on critical block removal

tellthemachines opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes #40618.

(edited by @mcsf:)

Show a modal asking the user for confirmation when attempting to remove Post Content and Query blocks, but only in the site editor.

~I have questions regarding implementation, mainly: should this initial version be extensible? If so, perhaps it would be better to register block criticality, as well as the message to display, in the block metadata. That would mean that third parties would be able to start using it immediately.~ For now, keep everything private using Gutenberg's new lock/unlock APIs. Everything is implemented inside the `@wordpress/block-editor` package. The `@wordpress/edit-site` just needs to unlock `BlockRemovalWarningModal` and render it in its `Editor` component.

~Another thought I had was should the check for whether to show the prompt be part of the `removeBlocks` action instead?~ Yes, this works inside the `removeBlocks` call. We've introduced some dedicated logic, introduced a private action named `privateRemoveBlocks`, and this new logic now comes with an extra argument, `forceRemove`, that is enabled when the user confirms deletion via the modal.


## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

A) In the site editor, go to a template that has either Post Content or Query (both considered critical blocks) and try the following:

1. Select the critical block and press backspace.
2. Select the critical block, navigate to its toolbar options dropdown and click "Delete".
3. Try both actions above with more than one block selected, including the critical block.
4. In the list view, try deleting the block with backspace, and through the options dropdown.
5. Also in the list view, try the above with more than one block selected, including the critical block.
6. In all these situations, the modal should pop up asking for confirmation of critical block removal.
~7. For bonus points: check if I missed any block removal methods 😄~

B) Make sure that there is no change in behaviour in other editors. For instance:
1. Open a new post in the Post Editor.
2. Insert one of the targeted blocks, e.g. by pasting in `<!-- wp:query /-->` 
3. Make sure you are able to delete it without seeing the modal.


### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

<img width="387" alt="51145-deleting-single" src="https://github.com/WordPress/gutenberg/assets/150562/42ff9669-cdcc-4342-a761-0b9d4947a3d6">
<img width="419" alt="51145-deleting-multiple" src="https://github.com/WordPress/gutenberg/assets/150562/beec51c0-2cf9-444e-9aa8-173b916b5076">

tellthemachines committed:
Show warning on critical block removal

ramonjd commented on packages/block-editor/src/components/block-actions/index.js:
```diff
@@ -74,7 +77,21 @@ export default function BlockActions( {
 		},
 		onRemove() {
-			return removeBlocks( clientIds, updateSelection );
+			const shouldDisplayRemovalPrompt = clientIds
```
Is this a possible candidate for memoization? I'm not sure, I guess `clientIds` would be constantly changing. It might be more optimal assuming a user is trying to remove the same block from the same block tree 😄

ramonjd commented on packages/block-editor/src/components/list-view/block-select-button.js:
```diff
@@ -105,7 +108,22 @@ function ListViewBlockSelectButton(
 				firstBlockRootClientId;
 
-			removeBlocks( blocksToDelete, false );
+			const shouldDisplayRemovalPrompt = blocksToDelete
```
Similar thought to the `shouldDisplayRemovalPrompt` above.

ramonjd commented on packages/block-editor/src/utils/show-block-removal-warning.js:
```diff
@@ -0,0 +1,55 @@
+	const message =
+		blockName === 'core/post-content'
+			? __(
+					'This block displays the content of a post or page. Removing it it is not advised.'
```
Should we indicate why it's not advised as mentioned in the [issue](https://github.com/WordPress/gutenberg/issues/40618#issuecomment-1273198648)?

I'm leaning towards "yes",  however I suppose that opens the can of worms that we'd have to customize the message for every warning depending on the block. 🤔

ramonjd commented on packages/block-editor/src/components/writing-flow/use-input.js:
```diff
@@ -66,7 +69,24 @@ export default function useInput() {
 				event.preventDefault();
 				if ( __unstableIsFullySelected() ) {
-					removeBlocks( getSelectedBlockClientIds() );
+					const shouldDisplayRemovalPrompt =
```
Maybe this should be something like `blockNamesWithWarnings` or something way better. `should*` makes me thing the return values is a bool

ramonjd commented on packages/block-editor/src/utils/show-block-removal-warning.js:
```diff
@@ -0,0 +1,55 @@
+} from '@wordpress/components';
+import { __, sprintf } from '@wordpress/i18n';
+
+export function showBlockRemovalWarning( blockName ) {
```
Hmm, I'm wondering about the mixed type return value. The function name, at least in my mind, implies a bool, but I see `blockName` is handy to have to display the removal prompt content.

Could we rather pass down a list of IDs and return those that required removal. That means we could memoize in one spot too. E.g.,

`getBlockNamesWithRemovalWarnings( clientIds ) // => [ 'core/query' ]`

```js
/**
 * 
 * @param {Array} clientIds 
 * @returns {Array}
 */
const getBlockNamesWithRemovalWarnings = ( clientIds ) =>
    clientIds
        .map( getBlockName )
        .filter( ( blockName ) => {
             return blockName === 'core/query' || blockName === 'core/post-content';
        } );
```

Would that work?

ramonjd reviewed (commented):
Thanks for working on this! I haven't tested manually yet, just left some flyby comments 😄

andrewserong commented on packages/block-editor/src/utils/show-block-removal-warning.js:
```diff
@@ -0,0 +1,55 @@
+			title={ sprintf(
+				/* translators: %s: the name of a menu to delete */
+				__( 'Remove %s?' ),
+				blockName
```
Can we grab the human friendly (display) name here? Maybe via `useBlockDisplayInformation`, since currently it shows up as `core/query`:

<img width="503" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/10236eda-d6f6-4ea7-8247-3b3489bb8423">

tellthemachines commented:
Thanks for the feedback @ramonjd ! Those are all great points, but I'd love to have more thoughts on the approach right now. Details can be ironed out once we're sure this is the way to go 😅

andrewserong reviewed (commented):
Nice work, it's testing pretty well!

One thing I noticed is that this applies in the post editor, too, so if you're editing a post or page that contains a Latest Posts section and go to remove the Query block, the warning will also display. Is that intended? Would it be worth further limiting the warning so that it only displays when editing a template or template part?

> This is a first approach: it works, but the code is messy.

In terms of the logic, I actually quite like it, and that it's explicit about when it's being executed, rather than on every call of `removeBlocks`. The main thing I was wondering from looking over the code is whether we can abstract the logic into a hook? That way, if it's only an extra couple of lines in each place it's used, I'm wondering if that might make it feel less messy when it has to be manually introduced in each place where we need it? 🤔

> should this initial version be extensible?

Personally, I quite like the hard-coded idea to begin with, since the idea of the feature is quite tied to the idea of what is required in a template in order for it to function. But I could imagine potentially having a property in `block.json` or something like that in the future, if folks wanted to have a flag to say "warn on removal", like you mention.

andrewserong commented:
> Details can be ironed out once we're sure this is the way to go 😅

Sounds good — apologies if I went too detailed, too!

andrewserong commented:
Just another thought for extensibility — I wonder if it'd be helpful storing the list of critical blocks to warn on removal in state somewhere? For example when core blocks are registered, `setGroupingBlockName` is called with the group block's name ([here](https://github.com/WordPress/gutenberg/blob/HEAD/packages/block-library/src/index.js#L270)), so potentially that could be a similar place to register a canonical list of critical blocks to warn on removal? For the moment, I'd probably lean toward sticking with a hard-coded list, as I imagine it'd be easier to make it extensible in a follow-up, than the other way around? 🤔

ramonjd commented:
> I have questions regarding implementation, mainly: should this initial version be extensible? 

Agree with @andrewserong, I think it'd be great to keep a smaller footprint and even make the store methods private for now until things are stabilized. 

> Another thought I had was should the check for whether to show the prompt be part of the removeBlocks action instead?

I agree with your instinct. Until there are more blocks on the warning list, I'm assuming it might be more optimal the way you've done it (?)

> Details can be ironed out once we're sure this is the way to go

You bet!

ndiego commented:
This is looking great, but I do want to call out the need for extensibility in the long run. I can imagine people wanting to add additional blocks to this "critical" list. I'm thinking about those that are building custom blocks for clients or perhaps e-commerce blocks. 

There also needs to be a way to turn these warnings off. I think this is probably more important than extensibility right now. While helpful to the average user, I can see these warnings becoming a bit annoying for power users or theme builders who are in the process of building out a block theme.

ramonjd commented:
> While helpful to the average user, I can see these warnings becoming a bit annoying for power users or theme builders who are in the process of building out a block theme.

Good point 😄 I think most admins would turn it off immediately, or soon after they become familiar with blocks.

I suppose the warning should furthermore only arrive on the scene for already-published blocks, or blocks with changes that have been saved once? Otherwise we see something like this:

<img width="757" alt="Screenshot 2023-06-02 at 10 44 28 am" src="https://github.com/WordPress/gutenberg/assets/6458278/d3a231eb-3eb3-4643-aec2-3796e79de125">

I'm also wondering whether a warning for the query block (for example) should be shown at all in the post editor 🤷 I guess making it extensible would allow site owners to build their own rules. 

Especially given that the premise behind the feature, that there are some blocks that one shouldn't remove, in my mind at least, is situational and could be unique to every site.

Less intrusive than a popup could be a notification that appears after the deletion of target blocks and offers and "Undo" step 🤷 

<img width="461" alt="Screenshot 2023-06-02 at 10 47 18 am" src="https://github.com/WordPress/gutenberg/assets/6458278/18e7833a-6b67-4749-bb97-b7915da64ece">

tellthemachines commented:
Thanks for all the thoughts folks ❤️ 

> One thing I noticed is that this applies in the post editor, too, so if you're editing a post or page that contains a Latest Posts section and go to remove the Query block, the warning will also display. Is that intended?

I didn't really consider it! It doesn't seem as necessary on posts or pages, because it's doubtful Query will be the main content there. I'll look into limiting the prompts to templates for now. That's also something we'll need to consider whether to make customisable or not.

> The main thing I was wondering from looking over the code is whether we can abstract the logic into a hook?

Something like `shouldDisplayRemovalPrompt(blocksToRemove, removalFunction)`, which takes care of looking up the block names and calling `displayRemovalPrompt`? That could work! (Function names tbd; the naming in this PR is all over the place, as @ramonjd observed already 😂 )

> apologies if I went too detailed, too!

No need to apologise! I just needed to work out the high-level shape of things before starting to address the finer points 😄 

> For the moment, I'd probably lean toward sticking with a hard-coded list, as I imagine it'd be easier to make it extensible in a follow-up, than the other way around?

Yeah, I think the extensibility bit will need some more thought. We'll likely need extenders to provide a custom message for the removal warning too. 

> Until there are more blocks on the warning list, I'm assuming it might be more optimal the way you've done it (?)

👍  It might not even be necessary to add the removal prompt logic anywhere else; that's probably something that can be worked out once there's a public API and we start seeing what kind of real world usage it gets.

> There also needs to be a way to turn these warnings off. I think this is probably more important than extensibility right now.

Good point! Let's get some testing feedback on the initial version and see how best to approach configurability in the UI. I'm thinking it could be something like a "Don't show again" checkbox in the modal itself.

tellthemachines commented:
> I suppose the warning should furthermore only arrive on the scene for already-published blocks, or blocks with changes that have been saved once?

The problem is we don't want to accidentally remove empty Post Content blocks - that's probably a more common occurrence than removing Query blocks. I think this is one of those things that we'll have to road test and smooth out iteratively.

ntsekouras commented:
Thanks for the PR @tellthemachines! 

>But I could imagine potentially having a property in block.json or something like that in the future, if folks wanted to have a flag to say "warn on removal", like you mention.

I definitely agree with @andrewserong about this. This seems like it should be a part of Blocks API and not hardcode some block names and messages in block editor package. This API in my mind would need to be aware of context(site editor, post editor, template, etc..) and provide different messages conditionally. Being a Block API means it will be also filterable with the existing filters for block settings.

tellthemachines commented:
> This seems like it should be a part of Blocks API and not hardcode some block names and messages in block editor package. 

Sure, the consensus is that there should be an API, but for the initial experiment it's better to just hardcode the block names so we can quickly get something out there to be tested, and get some feedback to help inform what kind of API is needed.

> This API in my mind would need to be aware of context(site editor, post editor, template, etc..) and provide different messages conditionally.

I've been thinking about how best to provide this context. If there is to be an API that allows extenders to define third party blocks as "critical" and provide a message for the prompt, it would make sense that extenders can also configure which editors the prompt displays in. The other thing to consider is if this API should also work with non-WP block editors, in which case we shouldn't hardcode editor context either.

tellthemachines committed:
Extract prompt display logic into a hook

tellthemachines commented:
I've now put all the logic that determines whether to show the prompt in a hook which returns a function that can be called instead of `replaceBlocks` with the same arguments.

Regarding the problem of only displaying the prompts in the site editor, I tried moving the modal into the site editor itself, around [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/block-editor/index.js#LL218C11-L218C11) but it's not working correctly. I'm not sure if there's a better place to put it in the site editor, or if there's another reason it doesn't work. Will continue looking into it tomorrow, but any ideas welcome in the meantime!

mcsf commented on packages/block-editor/src/components/block-list/use-block-props/use-selected-block-event-handlers.js:
```diff
@@ -20,14 +21,21 @@ import { store as blockEditorStore } from '../../../store';
 export function useEventHandlers( clientId ) {
-	const isSelected = useSelect(
-		( select ) => select( blockEditorStore ).isBlockSelected( clientId ),
+	const { isSelected } = useSelect(
+		( select ) => {
+			return {
+				isSelected:
+					select( blockEditorStore ).isBlockSelected( clientId ),
+			};
+		},
```
Why this change?

tellthemachines replied:
Oh, I should revert that! I had a couple more selectors in there but removed them with the latest commit.

mcsf commented on packages/block-editor/src/utils/show-block-removal-warning.js:
```diff
@@ -0,0 +1,103 @@
+		);
+		if ( firstCriticalBlock ) {
+			displayRemovalPrompt( true, {
+				removalFunction: () => {
+					removeBlocks( clientIds, selectPrevious );
+				},
```
Is the removal function ever anything other than `removeBlocks( clientIds, selectPrevious )`? Looking at the branch, it doesn't seem so. Seems like one piece of state that could be removed from `isRemovalPromptDisplayed` and one inch of API surface reduced.

tellthemachines replied:
It's always `removeBlocks`, the only difference is whether `selectPrevious` is true or false. I guess we could just pass `selectPrevious` instead of the whole function.

mcsf commented on packages/block-editor/src/utils/show-block-removal-warning.js:
```diff
@@ -0,0 +1,103 @@
+	};
+}
+
+export function BlockRemovalWarningModal( {
```
Should probably be under `../components`

mcsf commented on packages/block-editor/src/utils/show-block-removal-warning.js:
```diff
@@ -0,0 +1,103 @@
+	closeModal,
+	removalFunction,
+} ) {
+	const message =
```
The messages should sit next to the corresponding block types. Something like:

```js
const blockTypePromptMessages = new Map([
    [ 'core/query', __( 'This block ... ' ) ],
    [ 'core/post-content', __( 'This block ... ' ) ],
]);
```

to be shared by both hook and component.

(I know others have suggested making this pluggable, but I personally wouldn't rush it.)

tellthemachines replied:
Agreed, no hurry to create a public API for this.

mcsf commented on packages/block-editor/src/utils/show-block-removal-warning.js:
```diff
@@ -0,0 +1,103 @@
+				</Button>
+				<Button variant="primary" onClick={ () => onConfirmRemoval() }>
+					{ __( 'Confirm' ) }
+				</Button>
```
```suggestion
				<Button variant="tertiary" onClick={ closeModal }>
					{ __( 'Cancel' ) }
				</Button>
				<Button variant="primary" onClick={ onConfirmRemoval }>
					{ __( 'Confirm' ) }
				</Button>
```

mcsf commented on packages/block-editor/src/utils/show-block-removal-warning.js:
```diff
@@ -0,0 +1,103 @@
+	const { displayRemovalPrompt, removeBlocks } =
+		useDispatch( blockEditorStore );
+
+	function findFirstCriticalBlock( clientIds ) {
```
~This has the name of a `find` but the semantics of a `map`. Why not implement as `clientIds.find( ... )`?~

By stopping at the first matching _critical_* block, users can end up with the following bad UX:
* user clicks _Delete_ on a top-level group
* the group includes many nested block, among which a Query Loop and a Post Content 
* the prompt reads something like: "Remove Query Loop? This block does […]."
* user is OK with this choice and confirms, but user doesn't realise that Post Content is also up for deletion
* Post Content is unexpectedly deleted

*: I'm not a fan of the terminology, but we can discuss that separately.

mcsf replied:
Related: I experienced this when attempting to delete a Post Content block in the site editor :) :

<img width="601" alt="51145-remove-false-false" src="https://github.com/WordPress/gutenberg/assets/150562/4bfa0b3c-6013-4c20-84c8-910eaf5230c9">

tellthemachines replied:
Huh, thought I'd fixed that 😅 

Yeah it's not ideal. I guess we could show all the critical blocks to remove in the prompt; not sure what description to display in that case though.

tellthemachines replied:
I've updated this to show a list of comma-separated critical blocks (even if it's the same block repeated, though if it is repeated the message gets deduped). It doesn't look great but we have to allow for translatability of block names and messages so it's probably always going to be awkward to stick the pieces together 😬

mcsf replied:
Fixed and consolidated the algorithm in c972adf2993a2fe8625ccc33fc0f8edbcc6f3d3d

mcsf replied:
I took the liberty of trying something different in 6f236d2361085236224ee99580bc54d9daa10aeb. Please feel free to revert any of commits, as I don't mean to overstep! Screenshots for reference:

Single matching block | Multiple matching blocks
-|-
<img width="441" alt="Screenshot 2023-06-06 at 17 50 06" src="https://github.com/WordPress/gutenberg/assets/150562/fccc8303-f94a-49f1-8fad-78118f0bf772"> | <img width="446" alt="Screenshot 2023-06-06 at 17 50 12" src="https://github.com/WordPress/gutenberg/assets/150562/24a23c9b-18de-4248-8891-52ea99aa3bfb">

tellthemachines replied:
That's definitely better than what I had at first 😄 
I wonder if the block names should be bold or something, so it's more obvious at a glance which block(s) triggered the prompt. Might be good to get thoughts from design! cc @jameskoster 

Also, there was talk in #40618 about explaining _why_ removing these blocks is not recommended; if the block-specific message is expanded to include that reasoning perhaps the "removing these blocks..." bit won't be needed.

mcsf commented on packages/block-editor/src/utils/show-block-removal-warning.js:
```diff
@@ -0,0 +1,103 @@
+			if ( innerBlocks.length ) {
+				return findFirstCriticalBlock( innerBlocks );
+			}
+			return false;
```
Minor point, but following `find` convention, this could/should be `undefined`.

mcsf replied:
Renamed to `get*` in 5059a6730b4f762db349b6a44925693b7341c7b2

mcsf commented on packages/block-editor/src/utils/show-block-removal-warning.js:
```diff
@@ -0,0 +1,103 @@
+ */
+import { store as blockEditorStore } from '../store';
+
+function isBlockCritical( blockName ) {
```
See comment about `blockTypePromptMessages`

mcsf replied:
Removed in 82839d23ed105a611d1d65aefc0272d83a3ee62c

mcsf commented on packages/block-editor/src/components/block-list/index.js:
```diff
@@ -49,22 +50,37 @@ const pendingBlockVisibilityUpdatesPerRegistry = new WeakMap();
+			isOutlineMode: outlineMode,
+			isFocusMode: focusMode,
+			editorMode: __unstableGetEditorMode(),
+			displayPrompt: _displayPrompt,
+			removalFunction: _removalFunction,
+			blockToRemove: blockName,
```
See my other comment about the need for a removal function. If we could distill the data needs here, I think we could reduce this to a nicer:

```suggestion
			shouldDisplayBlockRemovalPrompt: !! blockToBeRemoved(),
```

Where `select( 'core/block-editor' ).blockToBeRemoved` is the successor to `isRemovalPromptDisplayed` and only returns the block name if there is one, or `null/undefined`.

mcsf replied:
Looking at this again, I wonder if this state should even be read/handled by `BlockList` at all. Shouldn't `BlockRemovalWarningModal` just handle everything internally? We have many such components:

```js
		<>
			<elementContext.Provider value={ element }>
				<IntersectionObserver.Provider value={ intersectionObserver }>
					<div { ...innerBlocksProps } />
				</IntersectionObserver.Provider>
			</elementContext.Provider>
			<BlockRemovalWarningModal />
		</>

```

where `BlockRemovalWarningModal` might become `BlockRemovalPromptListener` or anything that better conveys its conditional nature.

tellthemachines replied:
Yeah good point, it'll be cleaner if the modal itself gets what it needs from state. As it stands, the modal needs to know both the name of the block to be removed, and the function to call if the user confirms the action.

tellthemachines replied:
> Shouldn't BlockRemovalWarningModal just handle everything internally?

I've been playing with this a bit but can't quite get it to work, I think because if there's no state change in the block list when a block deletion is attempted, there's no reason to re-render the component so the modal never gets triggered when it's needed. Plus it renders a bunch of times when the block list first loads which is kind of annoying 😅

mcsf replied:
I gave this a go in f0247fe65a38229afa5c28ccfc6f4858bdc6a539. This works, right?

tellthemachines replied:
Yes! I was pretty much trying to do that, so not sure what went wrong 😕 probably had a typo somewhere. Thanks!

mcsf commented on packages/block-editor/src/store/actions.js:
```diff
@@ -1238,6 +1238,24 @@ export function removeBlock( clientId, selectPrevious ) {
+ *
+ * @return {Object} Action object.
+ */
+export function displayRemovalPrompt( displayPrompt, options = {} ) {
+	const { removalFunction, blockName } = options;
+	return {
+		type: 'PROMPT_REMOVAL',
+		displayPrompt,
+		removalFunction,
+		blockName,
+	};
+}
```
I wonder if we can / want to fold this logic into the [`removeBlocks` action](https://github.com/WordPress/gutenberg/blob/549ab383905d35b3062f4c76d856ce6937cb646a/packages/block-editor/src/store/actions.js#L1198C14-L1225), since that's the central code path for removal **and** it already is the site for checks and conditional dispatching:

- checks for `canRemoveBlocks` before proceeding
- dispatches `selectPreviousBlock` alongside the removal action if requested
- ensures correct focus by dispatching `ensureDefaultBlock` at the very end

I'm not sure of this idea, so I'd like to be challenged, but what about something like:

```diff
export const removeBlocks =
-	( clientIds, selectPrevious = true ) =>
+	( clientIds, selectPrevious = true, forceRemove = false ) =>
	( { select, dispatch } ) => {
		...
		if ( ! canRemoveBlocks ) {
			return;
		}

+		if ( ! forceRemove && needsConsentForRemoval( blockIds ) ) {
+			dispatch( { type: 'PROMPT_FOR_BLOCK_REMOVAL', blockIds } );
+			return;
+		}

		...
		dispatch( { type: 'REMOVE_BLOCKS', clientIds } );
+		dispatch( { type: 'PROMPT_FOR_BLOCK_REMOVAL', null } );
		...
	};
```

then the modal would dispatch a new remove action with `forceRemove: true`

tellthemachines replied:
Yeah, I thought about that (see my little monologue in the PR description 😄) but hesitate to do it because `removeBlocks` gets called in a lot of places (itself or via `removeBlock`) where the prompt isn't relevant, and that would mean going through the logic to check if there's a critical block in the list of blocks to remove every time. 

It does feel conceptually tidier to have everything together in removeBlocks though.

I'm also happy to hear more thoughts about this!

mcsf commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -1469,6 +1469,27 @@ export function isSelectionEnabled( state = true, action ) {
+ *
+ * @return {boolean} Updated state.
+ */
+export function isRemovalPromptDisplayed( state = false, action ) {
+	switch ( action.type ) {
+		case 'PROMPT_REMOVAL':
+			return {
+				displayPrompt: action.displayPrompt,
+				removalFunction: action.removalFunction,
+				blockName: action.blockName,
+			};
+	}
+
+	return state;
+}
```
I believe this could come down to something like

```js
case 'PROMPT_FOR_BLOCK_REMOVAL':
  return action.blockIds;
```

with the corresponding name change.

mcsf reviewed (commented):
Hi! I left a mishmash of inline comments that reflect my thought process as I reviewed. That means that the comments at the very end (chronologically) are probably the most important ones. I hope that the most important questions are apparent, but if it's confusing please let me know and I can try to rephrase or sum up my thoughts. :)

mcsf commented on packages/block-editor/src/utils/show-block-removal-warning.js:
```diff
@@ -0,0 +1,55 @@
+	return (
+		<Modal
+			title={ sprintf(
+				/* translators: %s: the name of a menu to delete */
```
```suggestion
				/* translators: %s: the title of a block to delete, e.g. "Post Content" */
```

tellthemachines committed:
Revert formatting change.

tellthemachines committed:
Prompt for removal of all the blocks

mcsf committed:
Move prompt state handling out of BlockList and into self

mcsf committed:
findCriticalBlocks: don't dismiss children of matching node

* Refactor by embracing `flatMap` semantics

* When `isBlockCritical( blockName )`, don't immediately return: there
  could be other matching block types within that node's inner blocks.

mcsf committed:
findCriticalBlocks -> getBlocksToPromptFor

The intention here was to drop the "find" prefix, which can mislead the
render into believing the function is meant to return a single match or
none.

mcsf committed:
Drop isBlockCritical()

mcsf committed:
Redesign removal modal

This is an attempt to reconcile a desire for clarity in the UI as to
what is about to be deleted (and why that matters) and the i18n
constraints that apply:

* Remove `__( 'Remove %s?' )` string, which -- although manageable --
  places an added burden on translators of more inflected languages.

* Trim the strings in `blockTypePromptMessages`. This allows us to stack
  all matching blocks' custom messages and finish off with a single
  "Removing it is not advised." paragraph.

tellthemachines committed:
Move prompt into site editor.

tellthemachines commented:
Going back to the problem of getting the prompt to only display in the site editor, I just pushed an update that moves the modal into the site editor, and dispatches and action from inside the modal, flagging that the modal is available. Then the hook that checks if the prompt needs to be displayed also checks if the modal is available, and if it isn't it just goes ahead and removes the blocks. 

This is so that the prompt doesn't display when a Query block is removed from the post editor, and it was the best way I could find of doing it without requiring the block editor package to be aware of its context (post or site editor). Thoughts for improvement or other ideas welcome!

Still to do:

* Try passing just the `selectPrevious` parameter for `removeBlocks` instead of passing the whole removal function to the modal;
* Move modal into `components` folder; maybe `useBlockRemovalWarning` should go into `hooks` too;
* Revisit the naming of all these things 😅 
* Should these new actions and selectors be made private for now?
* Add tests;
* Decide on definitive modal text.

jasmussen commented:
Screenshot of the delete modal:

<img width="1392" alt="delete modal" src="https://github.com/WordPress/gutenberg/assets/1204802/5e3fca53-60df-40af-8f67-a962651aa245">

Looks good, works well. I would change the verbiage to be a single paragraph that says:

> Really delete this block?
> **Query Loop** displays a list of posts. If you remove this block, the template might stop working.
> [Cancel] [Delete Query Loop]

Not a strong opinion, and not blocking, but to me it reads a bit more clearly. What do you think?

jameskoster commented:
> Decide on definitive modal text.

How about:

### Post Content
> **Remove Post Content**
> This block displays the content of a post / page. If you remove it then your posts / pages will appear empty to visitors.
> [Cancel] [Remove]

### Query Loop
> **Remove Query Loop**
> This block retrieves posts for display. If you remove it then the template will not display any posts.
> [Cancel] [Remove]

I don't know if it needs to hold up this PR, but the Query Loop could use a little more logic around when to display the confirmation modal. Ideally it only appears when you're editing an "Archive Page" as defined in the [template hierarchy](https://developer.wordpress.org/themes/basics/template-hierarchy/). At the moment it seems to appear regardless of which template / part you're editing.

jasmussen commented:
That works.

mcsf committed:
Reset removalPromptExists upon listener unmount

mcsf committed:
Let action removeBlocks handle prompts and confirmations

mcsf committed:
Add private action `privateRemoveBlocks` to hide extended interface

mcsf commented:
@tellthemachines: I spent some time on this today, exploring the ideas we discussed privately. I'm sort of happy with how it's working. Meanwhile, there is still some polishing to do (more stuff to move to private actions/selectors, maybe refactor/simplify the state, etc.).

Also: after giving it a try, I'm not so convinced about using/extending `notices` for this. This feature probably belongs entirely inside `block-editor`.

LMK what you think!

tellthemachines committed:
Fix unit tests

tellthemachines commented on docs/reference-guides/data/data-core-block-editor.md:
```diff
@@ -1089,6 +1089,18 @@ _Returns_
 
 -   `boolean`: Is navigation mode enabled.
 
+### isRemovalPromptDisplayed
```
Perhaps the new actions and selectors could be made private for now, at least until we create an API for configuring prompts on any block or editor.

mcsf replied:
Yes, let's make everything private.

tellthemachines commented on packages/block-editor/src/utils/block-removal.js:
```diff
@@ -0,0 +1,89 @@
+	),
+};
+
+export function BlockRemovalWarningModal() {
```
There was a discussion somewhere above about moving this into components. Should it also be made private?

tellthemachines reviewed (commented):
Thanks @mcsf , those changes look good! I added a fix for the removeBlocks unit tests. There were a couple items still on my to-do list for this PR which I've commented on below, and I'll try get around to them today or tomorrow.

andrewserong commented on packages/block-editor/src/store/actions.js:
```diff
@@ -1238,6 +1189,37 @@ export function removeBlock( clientId, selectPrevious ) {
+ * @param {boolean} exists Whether a prompt display mechanism exists.
+ * @return {Object} Action object.
+ */
+export function removalPromptExists( exists = true ) {
```
I'm wondering if we can rename this action to be more explicit about the action it performs? Currently the selector and the action have the same name, which could be confusing. Would something like `enableRemovalPrompt` or `enableRemovalPrompts` work as an action name? Or, `setRemovalPromptStatus` or something like that?

andrewserong commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1303,6 +1303,32 @@ export function isSelectionEnabled( state ) {
+ */
+export function removalPromptExists( state ) {
+	return state.removalPromptExists;
+}
```
Nit: in terms of naming `exists` made me think of whether or not a prompt is currently displayed. This could just be me, but I was wondering if `enabled` would be a clearer word to use? Then we could use an `is` prefix for something like `isRemovalPromptEnabled` as a selector name, or would that be too close to `isRemovalPromptDisplayed`? Another option could be to use a plural `prompts` for the action and selector for the removal prompts in general, and singular `prompt` for whether or not it is currently displayed.

andrewserong commented on packages/block-editor/src/utils/block-removal.js:
```diff
@@ -0,0 +1,89 @@
+			title={ _n(
+				'Really delete this block?',
+				'Really delete these blocks?',
+				blockNames.length
```
Using `blockNames.length` means that the plural text only shows for me if I've selected multiple _types_ of critical blocks.

If a selection contains more than one block, should we use the plural in all cases, since the `Confirm` action will result in multiple blocks being deleted?

At the very least, would it be good if selecting multiple query loop blocks resulted in the plural? Here's how it looks with two Query Loop blocks selected:

<img width="971" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/a570b421-aba1-4065-8ad4-cb57937de36d">

mcsf replied:
At the time I gave this way too much thought, trying to decide between `blockNames.length` and `clientIds.length`. 😅 But I didn't consider the fact that, after eventually turning `blockNames` into a deduped array, we could run into situations like the one you show. I'm fine with switching to `clientIds`, which incidentally per [your other comment](https://github.com/WordPress/gutenberg/pull/51145/files#r1236226593) is a piece of data we should add.

andrewserong commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1303,6 +1303,32 @@ export function isSelectionEnabled( state ) {
+ *
+ * @return {Object} Boolean indicating if prompt should be displayed and removal function.
+ */
+export function isRemovalPromptDisplayed( state ) {
```
The `is` prefix to this selector made me think the selector would return a `boolean`. Could we change this to use a `get` prefix, since it's returning an object?

Alternately, we could have two selectors, one that returns a `boolean` (if we need to just check if it should be displayed), and another that returns the whole object?

andrewserong commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1303,6 +1303,32 @@ export function isSelectionEnabled( state ) {
+ * @return {Object} Boolean indicating if prompt should be displayed and removal function.
+ */
+export function isRemovalPromptDisplayed( state ) {
+	return {
```
Does this need to return a new object each time? The state for `isRemovalPromptDisplayed` only has the below three keys, so I was wondering if it'd be enough to return `state.isRemovalPromptDisplayed`.

andrewserong commented on docs/reference-guides/data/data-core-block-editor.md:
```diff
@@ -1089,6 +1089,18 @@ _Returns_
 
+### isRemovalPromptDisplayed
+
+Selector that returns if the removal prompt needs to be displayed.
```
The selector returns an object, so its result is always truthy, which could be confusing.

mcsf replied:
Agreed, we should change/rename this.

andrewserong commented on packages/block-editor/src/utils/block-removal.js:
```diff
@@ -0,0 +1,89 @@
+	const closeModal = () => displayRemovalPrompt( false );
+
+	const onConfirmRemoval = () => {
+		removalFunction();
```
Apologies if I'm missing details here, but why is `removalFunction` required? Could we call `removeBlocks` (possibly with the force flag) with the selected clientIds instead of storing a `removalFunction` in state?

Edit: ah, wait, sometimes a user might be attempting to delete a block that is not selected (e.g. in the list view), so I think I see why `removalFunction` is needed now. Feel free to ignore this comment!

Edit-edit: unless we stored the `clientIds` to be removed in state, rather than the removal function? Just an idea, though 🙂

mcsf replied:
Yeah, this is something I raised a couple of weeks ago, and we could just store the client IDs and the `selectPrevious` boolean.

andrewserong commented on packages/block-editor/src/utils/block-removal.js:
```diff
@@ -0,0 +1,89 @@
+	const { displayRemovalPrompt, removalPromptExists } =
+		useDispatch( blockEditorStore );
+
+	// Signalling the removal prompt is in place.
+	useEffect( () => {
+		removalPromptExists( true );
+		return () => {
+			removalPromptExists( false );
+		};
+	}, [ removalPromptExists ] );
```
Why is this performed in a `useEffect` when the modal is rendered? Would it be better to set this status when the editor is initialized, e.g. here: https://github.com/WordPress/gutenberg/blob/aba94155dcad8202a4210ba5c2d31ce172bf25aa/packages/edit-site/src/index.js#L38

mcsf replied:
Good point 👀

mcsf replied:
I gave this a go in a5cce0a9ed18ab72c5f37a2df5b355eb7cfa53ad but decided against per my commit message. WDYT?

andrewserong replied:
The argument in your commit message sounds reasonable to me! I don't feel strongly about it, and you make a good point that keeping it in the `useEffect` means that the modal component is kind of encapsulated in that it's responsible for handling its own state and setup. Sounds good to leave as-is for now 👍

andrewserong commented on packages/block-editor/src/utils/block-removal.js:
```diff
@@ -0,0 +1,89 @@
+};
+
+export function BlockRemovalWarningModal() {
+	const { removalFunction, blocksToPromptFor: blockNames } = useSelect(
```
If the `blocksToPromptFor` attribute needs to be renamed here, should we go for a different name in state?

mcsf replied:
Maybe consolidate around `blockNamesForPrompt`, since the values in question are block names.

andrewserong reviewed (commented):
This is generally testing very well @tellthemachines and @mcsf! The main practical issue I ran into was just surrounding when the plural is displayed or not.

I've left a few code-quality notes surrounding the naming of the selectors and actions, and whether the `removalFunction` needs to be stored in state, apologies if there are a couple of double-ups in my comments! Some of the naming read a little bit awkwardly to me, so I was wondering if we can improve the clarity a little. Also, would it work to set the prompt status when the editor is initialised, instead of within the rendering of the modal?

I agree with the comment that it could be good to make the selectors and actions private for now, that could help a bit if we're still undecided on any of the naming, as things could potentially be renamed in follow-ups if need be.

Overall I'm liking the feature, though!

mcsf committed:
Try: Dispatch setRemovalPromptStatus in edit-site init

On one hand, `initializeEditor` is where we would typically set this
sort of setting, which is why this is worth trying out.

On the other hand, interfering with the `REMOVE_BLOCKS` action is a big
deal, so we should be absolutely sure that the prompt status in the
reducer is always accurate. The `useEffect` approach has the advantage
of keeping everything in one place: by rendering a single component like
BlockRemovalWarningModal, we guarantee that both the setting is correct
and the editor will render removal prompts.

mcsf committed:
Revert "Try: Dispatch setRemovalPromptStatus in edit-site init"

This reverts commit a5cce0a9ed18ab72c5f37a2df5b355eb7cfa53ad.

I'll repeat my reasoning from the parent commit:

"[...] interfering with the `REMOVE_BLOCKS` action is a big
deal, so we should be absolutely sure that the prompt status in the
reducer is always accurate. The `useEffect` approach has the advantage
of keeping everything in one place: by rendering a single component like
BlockRemovalWarningModal, we guarantee that both the setting is correct
and the editor will render removal prompts."

mcsf committed:
Make all actions & selectors private. Rename things.

* Addresses pull request feedback.

* See `FIXME` notes.

mcsf commented on packages/block-editor/src/store/private-actions.js:
```diff
@@ -105,3 +114,154 @@ export function unsetBlockEditingMode( clientId = '' ) {
+		// @see https://github.com/WordPress/gutenberg/pull/51145
+		if (
+			! forceRemove &&
+			// FIXME what's the best way to unlock a private selector in this
+			// context?
+			select( ( state ) => removalPromptExists( state.root ) )
```
@adamziel, what's the correct way to use the private selector in this context? 🙏 

Everyone else: don't hit me. 🙈

mcsf commented on packages/block-editor/src/store/test/actions.js:
```diff
@@ -614,10 +614,17 @@ describe( 'actions', () => {
-				getBlockRootClientId: () => undefined,
-				canRemoveBlocks: () => true,
-			};
+			// FIXME: Once the private action `privateRemoveBlocks` is properly
+			// unlocking the private selector `removalPromptExists`, this test
+			// should rewritten to do the right thing.
+			const select = Object.assign(
+				( selector ) =>
+					selector( { root: { removalPromptExists: false } } ),
+				{
```
🙈

mcsf committed:
Make BlockRemovalWarningModal private

tellthemachines reviewed (commented):
Thanks for the updates @mcsf ! Code's looking good and the new names make a lot more sense 🎉 
I can't approve this as the author 😅 but it would be good to merge within the next 24 hours, so it can make 16.1 RC. Also curious to know what @adamziel might have to say about the private selector access 👀 but not super concerned - if it's just a code quality issue we can improve it later, and if it turns out to cause any undesirable behaviour we can fix it as a bug during Beta.

talldan commented on packages/block-editor/src/components/block-removal-warning-modal/index.js:
```diff
@@ -0,0 +1,94 @@
+		closeModal();
+	};
+
+	return (
+		<Modal
+			title={ _n(
+				'Really delete this block?',
+				'Really delete these blocks?',
+				clientIds.length
+			) }
+			onRequestClose={ closeModal }
+		>
+			{ blockNamesForPrompt.length === 1 ? (
+				<p>{ blockTypePromptMessages[ blockNamesForPrompt[ 0 ] ] }</p>
+			) : (
+				<ul style={ { listStyleType: 'disc', paddingLeft: '1rem' } }>
+					{ blockNamesForPrompt.map( ( name ) => (
+						<li key={ name }>
+							{ blockTypePromptMessages[ name ] }
+						</li>
+					) ) }
+				</ul>
+			) }
+			<p>
+				{ _n(
+					'Removing this block is not advised.',
+					'Removing these blocks is not advised.',
+					blockNamesForPrompt.length
+				) }
+			</p>
+			<HStack justify="right">
+				<Button variant="tertiary" onClick={ closeModal }>
+					{ __( 'Cancel' ) }
+				</Button>
+				<Button variant="primary" onClick={ onConfirmRemoval }>
+					{ __( 'Confirm' ) }
+				</Button>
+			</HStack>
+		</Modal>
+	);
```
Not a blocker, but just noting that this is pretty similar to the `ConfirmDialog` component.

There are possibly some differences (the title?), so it could be a later refactor.

mcsf replied:
Thanks, good observation.

andrewserong reviewed (approved):
Thanks for the updates @mcsf! The plural in the modal title now reflects whether or not multiple blocks are selected, while the text within the body of the modal still refers to how many critical blocks are to be removed, so that's reading well to me 👍

The renaming is looking good so far. In a follow-up (not a blocker for landing this for 6.3), I think it'd still be good to rename `isRemovalPromptDisplayed` and its key in state, since it stores an object instead of a `boolean`. Now that the selectors and actions are private, though, this can easily happen in a follow-up, and isn't exposed publicly 👍

As @tellthemachines mentions, similarly with accessing the private selector via importing `removalPromptExists` directly, that appears to be a good thing to fix up in terms of code quality, but doesn't affect functionality, so seems like good follow-up territory to me, if we need to get this PR in sooner.

From smoke testing, this is still testing well with the prompt only showing up in the site editor with post content or query loop block removals, and the post editor is unaffected.

LGTM! ✨

mcsf committed:
Cleanup: Remove BlockList changes from branch

mcsf committed:
Tweak removal message for Query. Tweak comments.

jameskoster commented:
Apologies if this was discussed and dismissed during implementation, but would it be possible to update the modal title / message according to the suggestions in [this comment](https://github.com/WordPress/gutenberg/pull/51145#issuecomment-1580323751)? It's quite important for the user to understand _why_ removing these blocks is not advised. 

If it's not possible let me know and we can tweak the current copy to be more consistent with similar flows.

mcsf committed:
Split action into displayRemovalPrompt & clearRemovalPrompt

The action types have also been renamed: DISPLAY_REMOVAL_PROMPT,
CLEAR_REMOVAL_PROMPT.

Simplifies the function signatures and hopefully makes everything a
little bit clearer.

mcsf committed:
Rename setRemovalPromptStatus to toggleRemovalPromptSupport

Also: rename the action type, then sprinkle with comments.

mcsf committed:
Rename isRemovalPromptDisplayed to getRemovalPromptData

New reducer: removalPromptData
New selector: getRemovalPromptData

mcsf committed:
Add missing @return to displayRemovalPrompt

mcsf commented:
> Apologies if this was discussed and dismissed during implementation, but would it be possible to update the modal title / message according to the suggestions in [this comment](https://github.com/WordPress/gutenberg/pull/51145#issuecomment-1580323751)? It's quite important for the user to understand _why_ removing these blocks is not advised.

It doesn't work as well in terms of i18n (see explanation [here](https://github.com/WordPress/gutenberg/issues/48963#issuecomment-1549351710) and [further down](https://github.com/WordPress/gutenberg/issues/48963#issuecomment-1570661831)). In short, in many languages out there you'd need to inflect the block names to reflect their syntactic function (e.g. you'd get something like _"Remove Query-Loop'n?"_). This kind of inflection affects fewer locales, though, and those translators are more likely to know how to handle it. So I guess we can always decide to put the onus on translators to do the right thing if we give them enough context using `translators:` notes. I'm undecided. 😓

jameskoster commented:
Thanks for explaining @mcsf. Is it easier if we omit the block name? We could use a generic "Delete block?" modal title, then block-specific content:

### Post Content
> **Delete block?**
> This block displays the content of a post/page. If you remove it then your posts/pages will appear empty to visitors.
> [Cancel] [Delete]

### Query Loop
> **Delete block?**
> This block retrieves posts for display. If you remove it then the template will not display any posts.
> [Cancel] [Delete]

### Multi-block
> **Delete blocks?**
> These blocks are critical to the functionality of this template. Removing them is not advised.
> [Cancel] [Delete]

mcsf commented:
@jameskoster: thanks, that seems like a better solution for i18n. I'll have a go. Can you just confirm that I understood the following correctly?

* Up until now, we list all matching blocks' notes (e.g. "Query Loop displays a list of posts or pages.") followed by a single "Removing this/these block(s) is not advised.".
* With your proposal, we only show the individual block's note when only one block type was flagged. If multiple blocks are flagged, then we just show a generic "These blocks are critical to the functionality of this template. Removing them is not advised.".

mcsf commented:
@jameskoster: if a user is about to delete multiple blocks, but _only one of which is significant for the prompt_, what would you like the modal to be?

If the answer is the multi-block example, then it doesn't give context on which block is significant. For example, if I select a Cover block and a sibling Post Content, I will see "These blocks are critical to the functionality of this template. […]" and might not understand that it's because of the Post Content (maybe I wasn't even aware that I had it selected?).

If the answer is the single-block example, the same problem occurs, worsened by the fact that it sounds like the editor thinks only one block is about to be deleted ("Delete block? This block retrieves posts for display. […]")

jameskoster commented:
> Up until now, we list all matching blocks' notes (e.g. "Query Loop displays a list of posts or pages.") followed by a single "Removing this/these block(s) is not advised.".

Apologies, I didn't see the contextual detail. That'll teach me to actually check out the PR rather than just scanning the changes on github 🙈

With this detail I think the modal contents are okay, and I would just change the title to "Are you sure?", and the primary button label to "Delete".

> if a user is about to delete multiple blocks, but only one of which is significant for the prompt, what would you like the modal to be?

In these cases I think it's okay to display the contextual detail for the relevant block(s). So if you select Query Loop and a following Paragraph you'd see:

> Are you sure?
> Query Loop displays a list of posts or pages.
> Removing this block is not advised.
> [Cancel] [Delete]

Apologies for the mix up.

mcsf committed:
Tweak modal copy per feedback

mcsf commented:
Thanks, @jameskoster! While I have you, here are two screenshots of the PR per the latest commit:

<img width="801" alt="Screenshot 2023-06-22 at 16 58 30" src="https://github.com/WordPress/gutenberg/assets/150562/031fec2c-f405-4af8-8f6f-f333a9c26e88">
<img width="480" alt="Screenshot 2023-06-22 at 16 59 32" src="https://github.com/WordPress/gutenberg/assets/150562/4bdd4f0d-34ba-4c3d-b6f5-1bc0ed4f4c01">

Anything else you'd like to get in? :)

jameskoster commented:
That looks good to me, thank you! The list could use a little style tweak, but we can look into that separately.

mcsf commented:
(I've heavily edited the PR description to include the decisions we have come to, extend the testing instructions, provide new screenshots, etc.)

mcsf committed:
Turns out private selectors are attached to the thunk proxy!

Well that was easy. :)

mcsf committed:
Don't export the new reducers

mcsf commented on packages/block-editor/src/components/block-removal-warning-modal/index.js:
```diff
@@ -0,0 +1,94 @@
+import { store as blockEditorStore } from '../../store';
+import { unlock } from '../../lock-unlock';
+
+// In certain editing contexts, we'd like to prevent accidental removal of
+// important blocks. For example, in the site editor, the Query Loop block is
+// deemed important. In such cases, we'll ask the user for confirmation that
+// they intended to remove such block(s).
+//
+// @see https://github.com/WordPress/gutenberg/pull/51145
+export const blockTypePromptMessages = {
+	'core/query': __( 'Query Loop displays a list of posts or pages.' ),
+	'core/post-content': __(
+		'Post Content displays the content of a post or page.'
+	),
+};
```
Right now this entangles `block-editor` with `edit-site`, since these are rules that only apply to the site editor. I don't want to hold up this PR any longer, but a future improvement could be to let the consumer (the component rendering `BlockRemovalWarningModal`) provide their own rules. Maybe:

```js
<BlockRemovalWarningModal rules={ myPromptMessages } />
```

As you can tell, I'm undecided on the terminology: are these messages? Rules? Warnings?

mcsf replied:
-> #51841

mcsf committed:
Fix tests

mcsf commented:
Tests are failing across Gutenberg, potentially solved by https://github.com/WordPress/gutenberg/pull/51790. In the meantime, I'll just merge this one. Thanks, everyone!

(Edit: E2E tests were fine when run locally.)

mcsf commented on packages/block-editor/src/store/private-actions.js:
```diff
@@ -105,3 +113,188 @@ export function unsetBlockEditingMode( clientId = '' ) {
+		// @see https://github.com/WordPress/gutenberg/pull/51145
+		if (
+			! forceRemove &&
+			// FIXME: Without this existence check, the unit tests for
+			// `__experimentalDeleteReusableBlock` in
+			// `packages/reusable-blocks/src/store/test/actions.js` fail due to
+			// the fact that the `registry` object passed to the thunk actions
+			// doesn't include this private action. This needs to be
+			// investigated to understand whether it's a real smell or if it's
+			// because not all store code has been updated to accommodate
+			// private selectors.
+			select.isRemovalPromptSupported &&
+			select.isRemovalPromptSupported()
```
@jsnajdr: I quickly chatted with Adam about this. Is it possible that we are missing something in our Jest setup to accommodate private selectors?

jsnajdr replied:
It's not because of Jest, but because of the cumbersome way how we register the `block-editor` store, with the deprecated `registerStore` function, forced to use this way by the `persist` plugin.
```js
export const store = createReduxStore( STORE_NAME, {
	...storeConfig,
	persist: [ 'preferences' ],
} );

// We will be able to use the `register` function once we switch
// the "preferences" persistence to use the new preferences package.
const registeredStore = registerStore( STORE_NAME, {
	...storeConfig,
	persist: [ 'preferences' ],
} );
unlock( registeredStore ).registerPrivateActions( privateActions );
unlock( registeredStore ).registerPrivateSelectors( privateSelectors );
```
This code:
1. creates a `store` store descriptor for the `block-editor` store.
2. instantiates and registers the `block-editor` store, but using a different descriptor, the one created internally inside `registerStore`.
3. registers private actions and selectors to that one instance (!) that was just created.

The unit test that is failing is creating a new `registry`, and registers (instantiates) a `block-editor` store in this registry, using the `store` descriptor.

But nobody ever added the private actions and selectors to this descriptor!

The short-term solution is to register the private actions/selectors also to `store`. The long-term solution is to finish and merge #39632 (by @talldan), stop using the `persist` plugin, and stop using `registerStore` for `block-editor`.

mcsf replied:
Thanks very much for the explanation! I agree that merging #39632 should be the way forward. In the meantime, if you like, I opened https://github.com/WordPress/gutenberg/pull/52088 for the short-term solution.
